features.ts

```diff
@@ -0,0 +1,390 @@
+import * as t from "io-ts";
+import { isLeft } from "fp-ts/lib/Either.js";
+import { PathReporter } from "io-ts/lib/PathReporter.js";
+import { Response } from "express";
+import { Request as JwtRequest } from "express-jwt";
+import { FindCursor, ModifyResult, ObjectId, WithId } from "mongodb";
+
+import { State } from "./globals.js";
+import { sceneToJson } from "./scenes.js";
+
+export interface MongoSceneFeature {
+  scene_id: ObjectId; 
+  feature_time: Date;
+}
+
+export interface MongoSceneFeatureQueue {
+  scene_ids: ObjectId[];
+}
+
+export interface HydratedSceneFeature {
+  id?: ObjectId;
+  scene: Record<string, any>;
+  feature_time: Date;
+}
+
+function timeStrippedDate(date: Date): Date {
+  const day = new Date(date);
+  day.setHours(0, 0, 0, 0);
+  return day;
+}
+
+export async function getFeaturesForDate(state: State, date: Date): Promise<FindCursor<WithId<MongoSceneFeature>>> {
+  const day = timeStrippedDate(date);
+  const nextDay = new Date(day);
+  nextDay.setDate(nextDay.getDate() + 1);
+
+  return getFeaturesForRange(state, day, nextDay);
+}
+
+export async function getFeaturesForRange(state: State, startDate: Date, endDate: Date) {
+  return state.features.find({
+    "$and": [
+      { feature_time: { "$gte": startDate } },
+      { feature_time: { "$lt": endDate } }
+    ]
+  });
+}
+
+export async function tryPopFromFeatureQueue(state: State): Promise<ModifyResult<MongoSceneFeatureQueue>> {
```

Comment:
Is this function going to do what we want? As I understand it the `$pop` operator removes an item from the affected list, and that item just ... disappears. I believe the return value of the `findOneAndUpdate` is the affected document, after the pop has been applied. But for our "pop"-type operations, we care about what that item was. Am I missing something here?

Reply:
We discussed this the other day, but just to record for posterity: the [`findOneAndUpdate`](https://mongodb.github.io/node-mongodb-native/6.3/classes/Collection.html#findOneAndUpdate) method actually returns the original, pre-update document, unless you tell it otherwise. (In a Node context, that's done via setting the `returnDocument` field in the query [options](https://mongodb.github.io/node-mongodb-native/6.3/interfaces/FindOneAndUpdateOptions.html#returnDocument))